contracts/Pool.sol

```diff
@@ -529,3 +529,105 @@ library ReceiverWeightsImpl {
         }
     }
 }
+
+/// @notice A list of cycles and their deltas of amounts received by a proxy.
+/// For each cycle there are stored two deltas, one for the cycle itself
+/// and one for the cycle right after it.
+/// It reduces storage access for some usage scenarios.
+/// The cycle is described by its own entry and an entry for the previous cycle.
+/// Iterable and with random access.
+struct ProxyDeltas {
+    mapping(uint64 => ProxyDeltasImpl.ProxyDeltaStored) data;
+}
+
+/// @notice Helper methods for proxy deltas list.
+/// The list works optimally if after applying a series of changes it's iterated over.
+/// The list uses 2 words of storage per stored cycle.
+library ProxyDeltasImpl {
+    struct ProxyDeltaStored {
+        uint64 next;
+        // --- SLOT BOUNDARY
+        Pool.AmtDelta delta;
+    }
+
+    uint64 internal constant CYCLE_ROOT = 0;
+    uint64 internal constant CYCLE_UNINITIALIZED = 0;
+    uint64 internal constant CYCLE_END = type(uint64).max;
+
+    /// @notice Return the next non-zero, non-obsolete delta and its cycle.
+    /// The order is undefined, it may or may not be chronological.
+    /// Prunes all the fully zeroed or obsolete items found between the current and the next cycle.
+    /// Iterating over the whole list prunes all the zeroed and obsolete items.
+    /// @param current The previously returned cycle or CYCLE_ROOT to start iterating
+    /// @param finishedCycle The last finished cycle.
+    /// Data describing only the finished cycles is considered obsolete.
```

Comment:
Suggestion: I think the following would make the docs a bit clearer

```suggestion
    /// Entries for cycles before `finishedCycle` are considered obsolete.
```

Reply:
Fixed.